src/Npgsql/NpgsqlStream.cs

```diff
@@ -0,0 +1,329 @@
+            return SeekAsync(offset, origin, default, false).GetAwaiter().GetResult();
+        }
+
+        /// <inheritdoc />
+        public virtual Task<long> SeekAsync(long offset, SeekOrigin origin, CancellationToken cancellationToken)
+        {
+            ValidateArguments(origin);
+            CheckCanSeek();
+            if (cancellationToken.IsCancellationRequested)
+                return Task.FromCanceled<long>(cancellationToken);
+            using (NoSynchronizationContextScope.Enter())
+                return SeekAsync(offset, origin, cancellationToken, true);
+        }
+
+        /// <inheritdoc />
+        internal protected virtual Task<long> SeekAsync(long offset, SeekOrigin origin, CancellationToken cancellationToken, bool async)
+            => throw new NotImplementedException();
+
+        #endregion
+
+        #region Dispose
+
+        /// <inheritdoc />
+        internal protected bool IsDisposed { get; protected set; }
+
+        /// <inheritdoc />
+        internal protected virtual void CheckDisposed()
+        {
+            if (IsDisposed)
+                throw new ObjectDisposedException(GetType().FullName, "Object disposed");
+        }
+
+        /// <inheritdoc />
+        protected sealed override void Dispose(bool disposing)
+        {
+            if (IsDisposed || !disposing)
+                return;
+
+            DisposeAsync(false).GetAwaiter().GetResult();
+
+            IsDisposed = true;
+        }
+
+        /// <inheritdoc />
+#if !NET461 && !NETSTANDARD2_0
+        public override ValueTask DisposeAsync()
+#else
+        public ValueTask DisposeAsync()
+#endif
+        {
+            if (IsDisposed)
+                return default;
+
+            using (NoSynchronizationContextScope.Enter())
+                return DisposeAsync(true).Then(() =>
+                    IsDisposed = true
+                );
+        }
+
+        /// <inheritdoc />
+        internal protected virtual ValueTask DisposeAsync(bool async) => default;
+
+        #endregion
+
+        #region APM
+
+        /// <inheritdoc />
+        public sealed override IAsyncResult BeginRead(byte[] buffer, int offset, int count, AsyncCallback callback, object? state)
+            => ReadAsync(buffer, offset, count).AsApm(callback, state);
+
+        /// <inheritdoc />
+        public sealed override IAsyncResult BeginWrite(byte[] buffer, int offset, int count, AsyncCallback callback, object? state)
+            => WriteAsync(buffer, offset, count).AsApm(callback, state);
+
+        /// <inheritdoc />
+        public sealed override int EndRead(IAsyncResult asyncResult)
+            => ((Task<int>)asyncResult).GetAwaiter().GetResult();
+
+        /// <inheritdoc />
+        public sealed override void EndWrite(IAsyncResult asyncResult)
+            => ((Task)asyncResult).GetAwaiter().GetResult();
+
+        #endregion
+
+        #region Capabilities
+
+        /// <inheritdoc />
+        internal protected /*readonly*/ bool _canRead;
```

Comment:
Should be private protected (elsewhere as well).

Reply:
`private protected` is for classes in the same file, so the current modifier is correct.

Reply:
@YohDeadfall how do you mean? [According to the docs](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/private-protected):

> A private protected member is accessible by types derived from the containing class, but only within its containing assembly.

Reply:
Probably my knowledge is a bit outdated since I remember the restriction.